# [5주차 기본과제] 뉴스 기사 분류를 Gemma로 구현하기


In [88]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("google/gemma-2-2b-it")
model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-2b-it",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /Users/obov/.cache/huggingface/token
Login successful


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/636 [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [89]:
from datasets import load_dataset


news_ds = load_dataset("fancyzhx/ag_news")
news_ds

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 7600
    })
})

In [90]:
world_ds = news_ds["train"].filter(lambda data: data["label"] == 0)
world_ds["text"][:3]

['Venezuelans Vote Early in Referendum on Chavez Rule (Reuters) Reuters - Venezuelans turned out early\\and in large numbers on Sunday to vote in a historic referendum\\that will either remove left-wing President Hugo Chavez from\\office or give him a new mandate to govern for the next two\\years.',
 'S.Koreans Clash with Police on Iraq Troop Dispatch (Reuters) Reuters - South Korean police used water cannon in\\central Seoul Sunday to disperse at least 7,000 protesters\\urging the government to reverse a controversial decision to\\send more troops to Iraq.',
 "Palestinians in Israeli Jails Start Hunger Strike (Reuters) Reuters - Thousands of Palestinian\\prisoners in Israeli jails began a hunger strike for better\\conditions Sunday, but Israel's security minister said he\\didn't care if they starved to death."]

In [91]:
sports_ds = news_ds["train"].filter(lambda data: data["label"] == 1)
sports_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [92]:
business_ds = news_ds["train"].filter(lambda data: data["label"] == 2)
business_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [93]:
sci_tech_ds = news_ds["train"].filter(lambda data: data["label"] == 3)
sci_tech_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [94]:
news_ds["train"].features

{'text': Value(dtype='string', id=None),
 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}

In [95]:
label_map = news_ds["train"].features["label"].names

In [97]:
news_ds["train"].info

DatasetInfo(description='', citation='', homepage='', license='', features={'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}, post_processed=None, supervised_keys=None, builder_name='parquet', dataset_name='ag_news', config_name='default', version=0.0.0, splits={'train': SplitInfo(name='train', num_bytes=29832303, num_examples=120000, shard_lengths=None, dataset_name='ag_news'), 'test': SplitInfo(name='test', num_bytes=1880424, num_examples=7600, shard_lengths=None, dataset_name='ag_news')}, download_checksums={'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/train-00000-of-00001.parquet': {'num_bytes': 18585438, 'checksum': None}, 'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/test-00000-of-00001.parquet': {'num_bytes': 1234829, 'checksum': None}}, download_size=19820267, post_processing_size=None, dataset_size=31712727, size_in_bytes=51532994)

In [98]:
from pprint import pprint

pprint(news_ds["train"].info)

DatasetInfo(description='',
            citation='',
            homepage='',
            license='',
            features={'label': ClassLabel(names=['World',
                                                 'Sports',
                                                 'Business',
                                                 'Sci/Tech'],
                                          id=None),
                      'text': Value(dtype='string', id=None)},
            post_processed=None,
            supervised_keys=None,
            builder_name='parquet',
            dataset_name='ag_news',
            config_name='default',
            version=0.0.0,
            splits={'test': SplitInfo(name='test',
                                      num_bytes=1880424,
                                      num_examples=7600,
                                      shard_lengths=None,
                                      dataset_name='ag_news'),
                    'train': SplitInfo(name='train',
   

In [108]:
tokenizer("World", return_tensors="pt")["input_ids"].shape[-1]
tokenizer("World", return_tensors="pt")

{'input_ids': tensor([[    2, 10772]]), 'attention_mask': tensor([[1, 1]])}

In [100]:
label_sentences = [
    "Batman: This is mainly about the world, the one we’re sworn to protect.",
    "Joker: Why so serious? Clearly related to Sports news!",
    "Penguin: It’s all about Business, and trust me, I know how to play that game.",
    "Bane: This article mainly covers Sci/Tech.",
]

task_description = """
Task Description:
The scene unfolds with four iconic characters from the Batman universe seated around a table, each reading the same article. Batman, ever composed and logical, is the first to speak, delivering a clear and concise assessment of the piece. The Joker follows, making a mocking and unpredictable remark, turning the topic on its head. Two-Face, as always, relies on the flip of a coin to determine his take. Finally, Bane, with his focus on power and control, offers his perspective, emphasizing the article’s deeper implications.

Your task is simple: “Who explained the article the best?”.

Now, it’s time to choose—who described it best?

ARTICLE:"""

In [101]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.mps.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

In [102]:
news_ds["train"][1]["text"]

'Carlyle Looks Toward Commercial Aerospace (Reuters) Reuters - Private investment firm Carlyle Group,\\which has a reputation for making well-timed and occasionally\\controversial plays in the defense industry, has quietly placed\\its bets on another part of the market.'

In [103]:
news_ds["train"][1]["label"]

2

In [104]:
text_label = news_ds["train"][179]

probs = zero_shot_classification(
    text_label["text"],
    task_description,
    label_sentences,
)
print(probs)
import numpy as np

print(text_label["label"])
np.argmax(np.array(probs)).item()

[99.28125, 70.71484375, 158.25, 48.431640625]
3


2

In [86]:
def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_news_ds = news_ds.map(preprocess_function, batched=True)
tokenized_news_ds

Map:   0%|          | 0/120000 [00:00<?, ? examples/s]

Map:   0%|          | 0/7600 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 7600
    })
})

In [107]:
import numpy as np
from tqdm import tqdm
from colorama import Fore, Style, Back
import torch.nn.functional as F
import humanize
import math
import random


def calculate_nll(probs_softmax, label):
    return -torch.log(probs_softmax[label])


def get_confidence(probs_softmax):
    top_two_probs, _ = torch.topk(probs_softmax, 2)
    # 가장 높은 확률을 두 번째로 높은 확률로 나누어 비율을 계산
    return math.log10(top_two_probs[0] / top_two_probs[1])


total = 50
n_corrects = 0

# 0부터 7600 사이에서 겹치지 않게 50개의 숫자 뽑기
random_numbers = random.sample(range(7601), total)

for i in tqdm(range(total)):

    print(f"{Fore.YELLOW}----------------------------------{Fore.RESET}")
    text = tokenized_news_ds["test"][random_numbers[i]]["text"]
    label = tokenized_news_ds["test"][random_numbers[i]]["label"]
    probs = zero_shot_classification(
        text,
        task_description,
        label_sentences,
    )

    probs_tensor = torch.tensor(probs)
    probs_softmax = F.softmax(probs_tensor, dim=0)

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1
        tqdm.write(f"Index {i}: {Fore.GREEN}Success 🎉 {Style.RESET_ALL} ")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
    else:
        tqdm.write(f"Index {i}: {Fore.RED}Fail 🧐 {Style.RESET_ALL}")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
        print(f"Failed Query: {Back.LIGHTYELLOW_EX}{text}{Back.RESET}")
        label, pred = label_map[label], label_map[pred]
        print(
            f"Label : {Fore.CYAN} {label} {Fore.RESET} Pred : {Fore.MAGENTA} {pred} {Fore.RESET}"
        )
    current_accuracy = n_corrects / (i + 1)
    print(f"Current Accuracy: {Fore.BLUE}{current_accuracy:,.2%}{Fore.RESET}")

print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")
print(
    f"Total Correct Predictions: {Style.BRIGHT}{n_corrects} ({n_corrects/total:.2%}){Style.RESET_ALL}"
)
print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")

  0%|          | 0/50 [00:00<?, ?it/s]

----------------------------------


  2%|▏         | 1/50 [00:04<03:31,  4.32s/it]

Index 0: Success 🎉  
dataset index: 6293
Confidence: 24.29
NLL: -0.00
Current Accuracy: 100.00%
----------------------------------


  4%|▍         | 2/50 [00:06<02:40,  3.33s/it]

Index 1: Success 🎉  
dataset index: 1819
Confidence: 25.18
NLL: -0.00
Current Accuracy: 100.00%
----------------------------------


  6%|▌         | 3/50 [00:09<02:23,  3.05s/it]

Index 2: Success 🎉  
dataset index: 3875
Confidence: 28.67
NLL: -0.00
Current Accuracy: 100.00%
----------------------------------


  8%|▊         | 4/50 [00:12<02:08,  2.80s/it]

Index 3: Fail 🧐 
dataset index: 2371
Confidence: 27.65
NLL: 91.03
Failed Query: Track heads in the right direction RENAULT #39;S Formula One team boss Flavio Briatore paid the Shanghai International Circuit the greatest compliment when he said yesterday:  #39;It #39;s going to be difficult to beat this one.
Label :  Sports  Pred :  Business 
Current Accuracy: 75.00%
----------------------------------


 10%|█         | 5/50 [00:14<02:04,  2.76s/it]

Index 4: Success 🎉  
dataset index: 7349
Confidence: 27.11
NLL: -0.00
Current Accuracy: 80.00%
----------------------------------


 12%|█▏        | 6/50 [00:17<01:59,  2.72s/it]

Index 5: Fail 🧐 
dataset index: 149
Confidence: 25.68
NLL: 81.83
Failed Query: Greek sprinters arrive at IOC hearing ATHENS (Reuters) - Greek sprinters Costas Kenteris and Katerina Thanou have arrived at an Athens hotel for an International Olympic Committee (IOC) hearing into their missed doped tests, a saga that has shamed and angered the Olympic host ...
Label :  Sports  Pred :  Business 
Current Accuracy: 66.67%
----------------------------------


 14%|█▍        | 7/50 [00:19<01:53,  2.65s/it]

Index 6: Fail 🧐 
dataset index: 6337
Confidence: 25.83
NLL: 59.48
Failed Query: Asia ; Freed UN Hostages Say Humbled by Support in Ordeal The UN workers, who helped to run a presidential election won last month by US-backed incumbent Karzai, discussed their ordeal with him at his presidential palace in the morning.
Label :  World  Pred :  Business 
Current Accuracy: 57.14%
----------------------------------


 16%|█▌        | 8/50 [00:22<01:49,  2.60s/it]

Index 7: Fail 🧐 
dataset index: 5912
Confidence: 28.95
NLL: inf
Failed Query: Sony plans dual-DVD for Mac The DRX-710UL-T external DVD burner supports both FireWire 400 and USB 2.0. It ships with Roxio Toast 6 Lite. Double-layer support means users can burn up to 8.5GB of data on a single DVDR DL disc.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 50.00%
----------------------------------


 18%|█▊        | 9/50 [00:24<01:45,  2.58s/it]

Index 8: Fail 🧐 
dataset index: 2723
Confidence: 32.82
NLL: 94.61
Failed Query: Redskins Coach Bemoans Questionable Calls (AP) AP - Apparently even a Hall of Fame coach doesn't get a break from the officials.
Label :  Sports  Pred :  Business 
Current Accuracy: 44.44%
----------------------------------


 20%|██        | 10/50 [00:27<01:45,  2.63s/it]

Index 9: Fail 🧐 
dataset index: 4143
Confidence: 23.06
NLL: 53.11
Failed Query: Earthquake shook Yunnan, China BAOSHAN, China - A strong earthquake shook southwest China. The epicenter of magnitude 5 quakes on the Richter scale was located in the province of Yunnan not far from the city of Baoshan.
Label :  World  Pred :  Business 
Current Accuracy: 40.00%
----------------------------------


 22%|██▏       | 11/50 [00:30<01:38,  2.54s/it]

Index 10: Fail 🧐 
dataset index: 3435
Confidence: 28.23
NLL: inf
Failed Query: Microsoft, Sun, Intel push IT management via Web services (InfoWorld) InfoWorld - Microsoft and Sun Microsystems on Thursday are publishing a specification to leverage Web services for managing a broad range of IT systems including PCs and devices on a network.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 36.36%
----------------------------------


 24%|██▍       | 12/50 [00:32<01:38,  2.58s/it]

Index 11: Fail 🧐 
dataset index: 5660
Confidence: 29.13
NLL: 67.08
Failed Query: Grief, tears as mother, seven children killed in house fire mourned (Canadian Press) Canadian Press - ST. CATHARINES, Ont. (CP) - About 1,000 mourners filled a church Saturday for a funeral service for a mother and her seven children killed when fire tore through their century-old rural southwestern Ontario home.
Label :  World  Pred :  Business 
Current Accuracy: 33.33%
----------------------------------


 26%|██▌       | 13/50 [00:35<01:33,  2.54s/it]

Index 12: Fail 🧐 
dataset index: 6573
Confidence: 20.41
NLL: 63.41
Failed Query: Dillon enjoys a grand time Corey Dillon keeps piling up the rushing yards for the Patriots, but he could care less. What Dillon wants to pile up is wins. He #39;s doing that, too, in his 
Label :  Sports  Pred :  Business 
Current Accuracy: 30.77%
----------------------------------


 28%|██▊       | 14/50 [00:37<01:30,  2.50s/it]

Index 13: Fail 🧐 
dataset index: 4131
Confidence: 27.82
NLL: inf
Failed Query: Soy-Fertility Risk Downplayed (WebMD) Eating a diet rich in soy or taking soy supplements probably won #39;t harm a woman #39;s fertility, according to a new study.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 28.57%
----------------------------------


 30%|███       | 15/50 [00:39<01:25,  2.44s/it]

Index 14: Fail 🧐 
dataset index: 7164
Confidence: 27.31
NLL: 92.08
Failed Query: Cuts offered, issue remains tax or cap The union #39;s proposal to end the lockout, made Thursday at the league #39;s Canadian headquarters, calls for a tax that would penalize -- and perhaps deter -- high-end payrolls.
Label :  Sports  Pred :  Business 
Current Accuracy: 26.67%
----------------------------------


 32%|███▏      | 16/50 [00:42<01:25,  2.52s/it]

Index 15: Fail 🧐 
dataset index: 6251
Confidence: 25.22
NLL: 58.08
Failed Query: Arafat #39;s nephew not ruling out poison as cause of death While Yasser Arafat #39;s nephew says toxicology tests on his uncle show no poisons were found in his system, Arafat #39;s nephew isn #39;t ruling that out as a cause of death.
Label :  World  Pred :  Business 
Current Accuracy: 25.00%
----------------------------------


 34%|███▍      | 17/50 [00:45<01:23,  2.52s/it]

Index 16: Success 🎉  
dataset index: 405
Confidence: 28.44
NLL: -0.00
Current Accuracy: 29.41%
----------------------------------


 36%|███▌      | 18/50 [00:47<01:20,  2.53s/it]

Index 17: Success 🎉  
dataset index: 640
Confidence: 21.83
NLL: -0.00
Current Accuracy: 33.33%
----------------------------------


 38%|███▊      | 19/50 [00:50<01:17,  2.50s/it]

Index 18: Fail 🧐 
dataset index: 869
Confidence: 25.89
NLL: 59.61
Failed Query: Allawi Meets Militants, Pushes Amnesty Iraq's interim prime minister said that he had held private meetings with representatives of insurgent groups from Fallujah, Ramadi and Samarra to  persuade them to accept a government amnesty offer.
Label :  World  Pred :  Business 
Current Accuracy: 31.58%
----------------------------------


 40%|████      | 20/50 [00:52<01:11,  2.39s/it]

Index 19: Fail 🧐 
dataset index: 5243
Confidence: 25.07
NLL: 57.72
Failed Query: In Wake of Attacks, State of Emergency Declared in Iraq The Iraqi government declared a state of emergency for 60 days as U.S. and Iraqi forces prepared for an expected assault on rebels in Fallujah.
Label :  World  Pred :  Business 
Current Accuracy: 30.00%
----------------------------------


 42%|████▏     | 21/50 [00:54<01:10,  2.43s/it]

Index 20: Fail 🧐 
dataset index: 1709
Confidence: 30.26
NLL: inf
Failed Query: Novell #39;s Microsoft attack completes Linux conversion Novell Inc. has completed its conversion to Linux by launching an attack on Microsoft Corp., claiming that the company has stifled software innovation and that the market will abandon Microsoft Windows at some point in the future.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 28.57%
----------------------------------


 44%|████▍     | 22/50 [00:57<01:09,  2.47s/it]

Index 21: Fail 🧐 
dataset index: 3679
Confidence: 27.02
NLL: inf
Failed Query: Feds Make A Strike Against Spyware Companies WASHINGTON -- In what regulators are calling a first, the federal government has asked for a court order to shut down a spyware operation.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 27.27%
----------------------------------


 46%|████▌     | 23/50 [00:59<01:06,  2.46s/it]

Index 22: Success 🎉  
dataset index: 6947
Confidence: 29.42
NLL: -0.00
Current Accuracy: 30.43%
----------------------------------


 48%|████▊     | 24/50 [01:02<01:05,  2.54s/it]

Index 23: Success 🎉  
dataset index: 2857
Confidence: 29.49
NLL: -0.00
Current Accuracy: 33.33%
----------------------------------


 50%|█████     | 25/50 [01:05<01:04,  2.58s/it]

Index 24: Fail 🧐 
dataset index: 7093
Confidence: 30.58
NLL: inf
Failed Query: Chicken Genome Sheds New Light on Human DNA A new study states that 60 of the genes in chicken have close relations to human DNA. This may not comfort those who frequently eat the creature, but may ponder this the next time they order a batch of chicken wings.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 32.00%
----------------------------------


 52%|█████▏    | 26/50 [01:07<01:00,  2.54s/it]

Index 25: Fail 🧐 
dataset index: 5384
Confidence: 24.61
NLL: 56.68
Failed Query: The Bag of Aeolus  quot;Aeolus was keeper of the winds. He gave a bag of evil winds to Odysseus, instructing him to keep it closed while a good breeze wafted him home.
Label :  World  Pred :  Business 
Current Accuracy: 30.77%
----------------------------------


 54%|█████▍    | 27/50 [01:10<01:00,  2.64s/it]

Index 26: Fail 🧐 
dataset index: 4696
Confidence: 27.01
NLL: 87.00
Failed Query: Kezman #39;s first hammer blow MATEJA KEZMAN finally broke his scoring duck to settle this League Cup derby at Stamford Bridge today. Kezman missed a string of chances before firing in from Joe Cole #39;s through-ball in the 57th-minute to quash the Hammers #39; hopes of making the last 16.
Label :  Sports  Pred :  Business 
Current Accuracy: 29.63%
----------------------------------


 56%|█████▌    | 28/50 [01:13<00:59,  2.70s/it]

Index 27: Fail 🧐 
dataset index: 1767
Confidence: 23.37
NLL: 74.83
Failed Query: THE 35TH RYDER CUP Tiger #39;s Cup doesn #39;t run over with success Bloomfield Township, Mich. -- The Ryder Cup is upon us, and you know what that means: Time to dress up as Mrs. Doubtfire and taunt Colin Montgomerie from behind the ropes?
Label :  Sports  Pred :  Business 
Current Accuracy: 28.57%
----------------------------------


 58%|█████▊    | 29/50 [01:18<01:12,  3.44s/it]

Index 28: Success 🎉  
dataset index: 1248
Confidence: 25.72
NLL: -0.00
Current Accuracy: 31.03%
----------------------------------


 60%|██████    | 30/50 [01:20<01:03,  3.16s/it]

Index 29: Fail 🧐 
dataset index: 3955
Confidence: 26.66
NLL: 84.18
Failed Query: MLB Not Likely to Punish Steroid Users (AP) AP - For all the fuss over reported admissions of steroid use by Barry Bonds, Jason Giambi and Gary Sheffield, major league baseball probably won't discipline them.
Label :  Sports  Pred :  Business 
Current Accuracy: 30.00%
----------------------------------


 62%|██████▏   | 31/50 [01:23<00:55,  2.94s/it]

Index 30: Fail 🧐 
dataset index: 1944
Confidence: 17.77
NLL: 71.34
Failed Query:  #39;Emperor #39; Adriano has Inter under his rule One match into the Italian league season and (Emperor) Adriano already has Inter Milan under his rule. The Brazilian striker has scored six goals in Inter #39;s first four matches this season, including 
Label :  Sports  Pred :  Business 
Current Accuracy: 29.03%
----------------------------------


 64%|██████▍   | 32/50 [01:26<00:51,  2.85s/it]

Index 31: Fail 🧐 
dataset index: 4154
Confidence: 30.57
NLL: 70.39
Failed Query: Kerry Win Would Trigger Fierce Senate Race (AP) AP - If John Kerry is elected president, Massachusetts would end up with its first Senate vacancy in 20 years, triggering a springtime special election that could determine the balance of power in Congress' upper chamber.
Label :  World  Pred :  Business 
Current Accuracy: 28.12%
----------------------------------


 66%|██████▌   | 33/50 [01:28<00:46,  2.75s/it]

Index 32: Fail 🧐 
dataset index: 716
Confidence: 28.42
NLL: inf
Failed Query: RealNetworks Gets in Content Business (AP) AP - RealNetworks Inc. survived the dot-com collapse and an assault from Microsoft Corp. Now it's trying to remake itself into a provider of paid Internet content.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 27.27%
----------------------------------


 68%|██████▊   | 34/50 [01:30<00:41,  2.62s/it]

Index 33: Success 🎉  
dataset index: 2971
Confidence: 24.91
NLL: -0.00
Current Accuracy: 29.41%
----------------------------------


 70%|███████   | 35/50 [01:33<00:41,  2.74s/it]

Index 34: Success 🎉  
dataset index: 3658
Confidence: 29.76
NLL: -0.00
Current Accuracy: 31.43%
----------------------------------


 72%|███████▏  | 36/50 [01:36<00:37,  2.67s/it]

Index 35: Fail 🧐 
dataset index: 692
Confidence: 30.17
NLL: 88.35
Failed Query: AL Wrap: Oakland's Durazo Piles on Misery for Baltimore  NEW YORK (Reuters) - Erubiel Durazo's three-run homer in  the second inning helped the Oakland Athletics remain top of  the American League (AL) West with a 9-4 win over the reeling  Baltimore Orioles Thursday.
Label :  Sports  Pred :  Business 
Current Accuracy: 30.56%
----------------------------------


 74%|███████▍  | 37/50 [01:38<00:34,  2.64s/it]

Index 36: Fail 🧐 
dataset index: 1688
Confidence: 26.23
NLL: 102.18
Failed Query: Help wanted by IT services firms The growing services industry is hiring, but tech workers looking for a job may need to do more than brush up on their coding.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 29.73%
----------------------------------


 76%|███████▌  | 38/50 [01:41<00:31,  2.63s/it]

Index 37: Fail 🧐 
dataset index: 5048
Confidence: 24.37
NLL: inf
Failed Query: Coming to Project Managers: Multiscreen Microsoft PC Microsoft Corp. has launched a new entry in its ongoing effort to bring more innovative PC form factors to marketin the somewhat quirky form of a high-end system specialized for project managers.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 28.95%
----------------------------------


 78%|███████▊  | 39/50 [01:44<00:29,  2.67s/it]

Index 38: Fail 🧐 
dataset index: 4145
Confidence: 26.08
NLL: inf
Failed Query: Online gambling attracts surfers More than four million Britons are regular internet gamblers according to new research.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 28.21%
----------------------------------


 80%|████████  | 40/50 [01:47<00:27,  2.72s/it]

Index 39: Success 🎉  
dataset index: 535
Confidence: 29.26
NLL: -0.00
Current Accuracy: 30.00%
----------------------------------


 82%|████████▏ | 41/50 [01:49<00:24,  2.69s/it]

Index 40: Fail 🧐 
dataset index: 2093
Confidence: 27.55
NLL: inf
Failed Query: Media center at your fingertips Apple #39;s splashy digital music player has emboldened Microsoft and other technology titans to move quickly to the next frontier in portable entertainment: the video iPod, so to speak.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 29.27%
----------------------------------


 84%|████████▍ | 42/50 [01:52<00:21,  2.68s/it]

Index 41: Success 🎉  
dataset index: 867
Confidence: 27.98
NLL: -0.00
Current Accuracy: 30.95%
----------------------------------


 86%|████████▌ | 43/50 [01:55<00:18,  2.65s/it]

Index 42: Success 🎉  
dataset index: 4164
Confidence: 31.39
NLL: -0.00
Current Accuracy: 32.56%
----------------------------------


 88%|████████▊ | 44/50 [01:57<00:15,  2.62s/it]

Index 43: Fail 🧐 
dataset index: 355
Confidence: 26.61
NLL: 61.27
Failed Query: Greek weightlifter awaits verdict Greek weightlifter Leonidas Sampanis will find out on Sunday if he is to be stripped of his medal.
Label :  World  Pred :  Business 
Current Accuracy: 31.82%
----------------------------------


 90%|█████████ | 45/50 [02:00<00:12,  2.57s/it]

Index 44: Success 🎉  
dataset index: 6243
Confidence: 23.74
NLL: -0.00
Current Accuracy: 33.33%
----------------------------------


 92%|█████████▏| 46/50 [02:02<00:10,  2.66s/it]

Index 45: Fail 🧐 
dataset index: 6699
Confidence: 26.30
NLL: 60.55
Failed Query: Cause of Indonesian Plane Crash Probed (AP) AP - Investigators picked through the wreckage of an Indonesian passenger plane that crashed in stormy weather, killing at least 32 people in the county's worst air accident in six years.
Label :  World  Pred :  Business 
Current Accuracy: 32.61%
----------------------------------


 94%|█████████▍| 47/50 [02:16<00:17,  5.99s/it]

Index 46: Success 🎉  
dataset index: 5129
Confidence: 25.58
NLL: -0.00
Current Accuracy: 34.04%
----------------------------------


 96%|█████████▌| 48/50 [02:19<00:09,  4.90s/it]

Index 47: Fail 🧐 
dataset index: 7022
Confidence: 23.85
NLL: 54.93
Failed Query: One billion  #39;denied a childhood #39; More than one billion children around the world face a brutal existence because of poverty, war and Aids, the UN children #39;s agency reports.
Label :  World  Pred :  Business 
Current Accuracy: 33.33%
----------------------------------


 98%|█████████▊| 49/50 [02:21<00:04,  4.20s/it]

Index 48: Fail 🧐 
dataset index: 7172
Confidence: 25.16
NLL: 57.94
Failed Query: Australian man killed by shark on Great Barrier Reef (Canadian Press) Canadian Press - CAIRNS, Australia (AP) - A 38-year-old Australian man bled to death Saturday after he was rescued from the jaws of a shark while spearfishing on the Great Barrier Reef, authorities said.
Label :  World  Pred :  Business 
Current Accuracy: 32.65%
----------------------------------


100%|██████████| 50/50 [02:24<00:00,  2.89s/it]

Index 49: Fail 🧐 
dataset index: 961
Confidence: 27.33
NLL: 62.92
Failed Query: Conditions Worsen in Darfur, U.N. Agencies Say (Reuters) Reuters - Conditions for 1.2 million Sudanese\displaced in Darfur continue to worsen amid violent attacks,\spreading disease, and heavy rains which wreak havoc with aid\convoys, United Nations agencies said on Tuesday.
Label :  World  Pred :  Business 
Current Accuracy: 32.00%
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Total Correct Predictions: 16 (32.00%)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>


---


# 실습: Zero-shot Classification

이번 실습에서는 open LLM을 가지고 zero-shot classification을 해봅니다. 먼저 필요한 library들을 설치합시다.


그 다음 Gemma-2B를 사용하기 위해 다음과 같은 작업을 진행합니다:

1. huggingface.co 계정 만들고 로그인하기
2. https://www.kaggle.com/models/google/gemma/license/consent 에서 Gemma license 동의하기
3. 홈 화면으로 돌아와, `Profile > Settings > Access Tokens` 메뉴로 들어와 "Write" type의 token 생성하기
4. 생성한 토큰을 아래 "HF TOKEN"에 불여넣고 셀을 실행하기.


In [ ]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")

이번에는 Gemma-2B를 가지고 간단한 text 생성을 해봅시다.
"What is your name?" 이라는 text를 넣었을 때 어떤 text가 생성되는지 살펴봅시다.


In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
input_ids

In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
outputs = model.generate(**input_ids)
print(tokenizer.decode(outputs[0]))

In [ ]:
tokens = input_ids["input_ids"]
print(tokens)
input_ids

2B의 작은 LLM이라 질좋은 답변이 나오지 않는 것을 알 수 있습니다.
이번에는 입력으로 넣어준 token들의 logit을 계산해봅시다.


In [ ]:
tokens = input_ids["input_ids"]
print(tokens)

logits = model(**input_ids).logits
for i in range(tokens.shape[-1]):
    token = tokens[0, i].item()
    print(logits[0, i, token])

위와 같이 모델 출력의 `.logits`을 통해 token들의 logit을 알 수 있습니다.
Logit은 높을 수록 token이 나올 확률이 높다는 뜻입니다.

이번에는 logit 계산을 통해 zero-shot classification을 구현해보도록 하겠습니다.


In [8]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.cuda.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

아래는 실제로 zero-shot classification을 해본 결과입니다.


In [ ]:
probs = zero_shot_classification(
    "I am happy!", "Is the sentence positive or negative?: ", ["positive", "negative"]
)
print(probs)

보시다시피 우리는 Gemma를 별도로 학습하지 않았음에도 불구하고 주어진 문장이 긍정적이라는 것을 정확하게 예측하고 있습니다.

다음은 영화 리뷰 감정 분석 task에 적용해봅시다.
먼저 data를 불러옵니다.


In [ ]:
from datasets import load_dataset


imdb = load_dataset("imdb")


def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_imdb = imdb.map(preprocess_function, batched=True)

그리고 `test` data에서 50개의 영화 리뷰에 대해 예측하는 코드는 다음과 같습니다.


In [ ]:
import numpy as np
from tqdm import tqdm


n_corrects = 0
for i in tqdm(range(50)):
    text = tokenized_imdb["test"][i]["text"]
    label = tokenized_imdb["test"][i]["label"]
    probs = zero_shot_classification(
        text,
        "A movie review is given. Decide that the movie review is positive or negative: ",
        labels=["Answer: negative.", "Answer: positive."],
    )

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1

print(n_corrects)

In [ ]:
text = tokenized_imdb["test"][i]["text"]
text

In [ ]:
label = tokenized_imdb["test"][i]["label"]
label

보시다시피 정확도 88%로, 매우 높은 성능을 보이는 것을 알 수 있습니다.
